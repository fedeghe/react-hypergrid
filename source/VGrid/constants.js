import React from 'react';
// the name of the component
export const CMPNAME = 'react-vgrid',
    // how many lines extra viewport up and down before virtualization ? 
    GAP = 3,

    // grid sizes
    WIDTH = 1200,
    HEIGHT = 800,
    
    // grid item size
    ITEM_HEIGHT = 150,
    ITEM_WIDTH = 200,
    
    // group component height 
    GROUP_COMPONENT_HEIGHT = 20,

    // id appended string
    RVG_ID = '_ID',

    // debouning values
    DEBOUNCE_SCROLLING = 50,
    DEBOUNCE_FILTERING = 50,

    DEFAULT_LOADER = () => (<div>loading</div>),

    // no data message
    NO_FILTER_DATA_MESSAGE = 'no data',

    // ungrouped group label
    UNGROUPED_LABEL = 'ungrouped',

    // unfilter families
    FILTERS = {
        ALL : 'ALL',
        GLOBAL : 'GLOBAL',
        FIELDS : 'FIELDS',
    },
    UIE = 'data-uie';
