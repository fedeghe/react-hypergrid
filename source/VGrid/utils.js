let count = 0;
const prefix = 'RVG_',
    isFunction = f => typeof f === 'function',
    debounce = (func, wait) => {
        let timeout,
            enabled = true;
        return (...params) => {
            clearTimeout(timeout);
            timeout = setTimeout(() => {
                if (!enabled) return;
                func(...params);
                enabled = false;
                setTimeout(() => enabled = true, wait);
            }, wait);
        };
    },
    escapeComma = r => `${r}`.replace(/,/g, '\\,'),
    removeID = (jsonData, rhgID) => jsonData.map(row => {
        var r = {...row};
        delete r[rhgID];
        return r;
    }),
    asXsv = (columns, jsonData, rhgID, separator) => {
        const lines = [],
            keys = columns.map(c => c.key);
        lines.push(keys.join(separator));
        removeID(jsonData, rhgID).forEach(row => {
            lines.push(keys.map(k => escapeComma(row[k])).join(separator));
        });
        return lines.join("\n");
    },
    asJson = removeID,
    trakTime = ({ what, time, opts }) =>
        console.info(`%c${opts.lib.toUpperCase()} 🐢 ${what} spent ${time}ms`, 'color:DodgerBlue'),
    doWarn = ({ message, opts }) =>
        console.warn(`${opts.lib.toUpperCase()} 🙉 ${message}`),
    doThrow = ({ message, opts }) => {
        throw `${opts.lib.toUpperCase()} 🚨 ${message}`;
    },
    throwIf = ({ condition, message, opts }) => {
        if (condition) {
            throw `${opts.lib.toUpperCase()} 🚨 ${message}`;
        }
    },
    uniqueID = {
        toString: () => {
            count += 1;
            return prefix + count;
        }
    };
    

export {
    isFunction,
    debounce,
    asXsv,
    asJson,
    trakTime, doWarn, doThrow, throwIf, uniqueID
};
