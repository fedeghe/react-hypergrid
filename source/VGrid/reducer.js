import React from 'react';
import { isFunction } from './utils';
let count = 0;
const prefix = 'HYG_',
    uniqueID = {
        toString: () => {
            count += 1;
            return prefix + count;
        }
    },

    __cleanFilters = _filters => Object.keys(_filters).reduce((acc, k) => {
        acc[k] = {
            filter: _filters[k].filter,
            value: ''
        };
        return acc;
    }, {}),

    __getVirtual = ({ scrollTop, dimensions, size, lineGap }) => {
        const { height, itemHeight, width, itemWidth } = dimensions,
            columns = Math.floor(width / itemWidth),
            lines = Math.ceil(size / columns),
            carpetHeight = lines * itemHeight,
            trigger = scrollTop > (lineGap + 1) * itemHeight,

            topLinesSkipped = Math.max(0, Math.floor(scrollTop / itemHeight) - lineGap),
            topFillerHeight = topLinesSkipped * itemHeight,
            linesToRender = 2 * lineGap + Math.ceil(height / itemHeight),
            dataHeight = linesToRender * itemHeight,
            maxRenderedItems = columns * linesToRender,
            bottomFillerHeight = Math.max(carpetHeight - topFillerHeight - dataHeight, 0),
            fromItem = trigger
                ? topLinesSkipped * columns
                : 0,
            toItem = trigger ? fromItem + maxRenderedItems : linesToRender * columns;
        return {
            fromItem,
            toItem,
            carpetHeight,
            topFillerHeight,
            bottomFillerHeight,
            linesToRender,
            dataHeight,
            maxRenderedItems,
            loading: false,
            lines,
            columns,
            scrollTop,
        };
    },

    __getDoFilter = ({fields, filters}) => {
        const {funcFilteredFields, valueFilteredFields} = fields.reduce((acc, f) => {
            acc[(f in filters)? 'funcFilteredFields' : 'valueFilteredFields'].push(f);
            return acc;
        }, {funcFilteredFields: [], valueFilteredFields: []});
        return  v => row => 
            funcFilteredFields[v ? 'some' : 'every'](fk => 
                filters[fk].filter({
                    userValue: v || filters[fk].value,
                    row
                })
            )
            ||
            valueFilteredFields.some(f => `${row[f]}`.includes(v));
    },

    reducer = (oldState, action) => {
        const { payload = {}, type } = action,
            {
                dimensions,
                originalData,
                filteredData,
                virtual,
                virtual: {
                    lineGap,
                    scrollTop
                },
                globalFilterValue,
                filters,
                fields,
                // globalFilterValue
            } = oldState,

            actions = {
                loading: () => ({
                    virtual: {
                        ...virtual,
                        loading: true
                    }
                }),

                filter: () => {
                    const { value, field } = payload,
                        // must start from everything
                        base = [...originalData],
                        isGlobalSearch = !field,
                        ret = {};
                    let _newData = base,
                        _globalFilterValue = globalFilterValue,
                        _newFilters = {...filters};

                    // first maybe update filter value
                    if (!isGlobalSearch && field in _newFilters) {
                        _newFilters = {
                            ..._newFilters,
                            [field]: {
                                ..._newFilters[field],
                                value
                            }
                        };
                    }
                    // eslint-disable-next-line one-var
                    const doFilter = __getDoFilter({fields, filters: _newFilters});
                        
                    // global ? 
                    if (isGlobalSearch || _globalFilterValue) {
                        if(isGlobalSearch) {
                            ret.globalFilterValue = value;
                            _globalFilterValue = value;
                        }
                        
                        _newData = _newData.filter(doFilter(_globalFilterValue));
                    }
                    //then use all available filters on their value (that is updated)
                    _newData = _newData.filter(doFilter());
                    
                    // eslint-disable-next-line one-var
                    const newVirtual = __getVirtual({ dimensions, size: _newData.length, scrollTop, lineGap }),
                        { fromItem, toItem } = newVirtual;
                    return {
                        ...ret,
                        data: _newData.slice(fromItem, toItem),
                        filters: _newFilters,
                        filteredData: _newData,
                        filtered: _newData.length,
                        virtual: {
                            ...virtual,
                            ...newVirtual,
                            scrollTop: 0
                        }
                    };
                },

                unFilterFields: () => {
                    let _globalFilterValue = globalFilterValue,
                        _newFilters = {...filters},
                        _filteredData = [...originalData];

                    const filteringFields = payload.filter(f => fields.includes(f));
                    
                    filteringFields.forEach(f => {_newFilters[f].value = '';});
                    
                    // eslint-disable-next-line one-var
                    const doFilter = __getDoFilter({
                        fields,
                        filters: _newFilters
                    });

                    _filteredData = _filteredData.filter(doFilter());

                    if (_globalFilterValue) {
                        _filteredData = _filteredData.filter(doFilter(_globalFilterValue));
                    }
                    
                    // eslint-disable-next-line one-var
                    const newVirtual = __getVirtual({
                            dimensions,
                            size: _filteredData.length,
                            scrollTop, lineGap
                        }),
                        { fromItem, toItem } = newVirtual;
                        
                    return {
                        filters: _newFilters,

                        filteredData: _filteredData,
                        data: _filteredData.slice(fromItem, toItem),
                        globalFilterValue: _globalFilterValue,
                        virtual: {
                            ...virtual,
                            ...newVirtual,
                        },
                        filtered: _filteredData.length
                    };
                },
                unFilter: () => {                    
                    let _globalFilterValue = globalFilterValue,
                        _newFilters = {...filters},
                        _filteredData = [...originalData];

                    const doFilter = __getDoFilter ({fields, filters: _newFilters});
                            
                    switch (payload) {
                        case '_ALL_':
                            _globalFilterValue = '';
                            _newFilters = __cleanFilters(filters);
                            break;
                        case '_GLOBAL_':
                            _globalFilterValue = '';
                            _filteredData = _filteredData.filter(doFilter());
                            break;
                        case '_FIELDS_':
                            _newFilters = __cleanFilters(filters);
                            _filteredData = _filteredData.filter(doFilter(_globalFilterValue));
                            break;
                    }
                    

                    // eslint-disable-next-line one-var
                    const newVirtual = __getVirtual({
                            dimensions,
                            size: _filteredData.length,
                            scrollTop, lineGap
                        }),
                        { fromItem, toItem } = newVirtual;
                        
                    return {
                        filters: _newFilters,

                        filteredData: _filteredData,
                        data: _filteredData.slice(fromItem, toItem),
                        globalFilterValue: _globalFilterValue,
                        virtual: {
                            ...virtual,
                            ...newVirtual,
                        },
                        filtered: _filteredData.length
                    };

                },
                scroll: () => {
                    const scrollTop = parseInt(payload, 10),
                        newVirtual = __getVirtual({ dimensions, size: filteredData.length, scrollTop, lineGap }),
                        { fromItem, toItem } = newVirtual;

                    return {
                        data: filteredData.slice(fromItem, toItem),
                        virtual: {
                            ...virtual,
                            ...newVirtual,
                        }
                    };
                }
            };

        if (type in actions)
            return {
                ...oldState,
                ...actions[type]()
            };
        return oldState;
    },
    init = (cnf = {}) => {
        const {
                data = [],
                lineGap = 2,
                Loader = () => (<div>loading</div>),
                dimensions: {
                    width = 1200,
                    height = 800,
                    itemHeight = 150,
                    itemWidth = 200
                } = {},
                rhgID = '_ID',
                debounceTimes: {
                    scrolling = 50,
                    filtering = 50,
                } = {},
                
                header: {
                    caption: {
                        Component: HeaderCaptionComponent = null,
                        height: headerCaptionHeight = 0
                    }
                } = {},
                footer: {
                    caption: {        
                        Component: FooterCaptionComponent = null,
                        height: footerCaptionHeight = 0
                    }
                } = {},

                events: {
                    onItemEnter,
                    onItemLeave,
                    onItemClick,
                } = {},
                headers = {},
                globalPreFilter = '',
                NoFilterData = () => 'no data',
                cls: {
                    HeaderCaption: HeaderCaptionCls = null,
                    FooterCaption: FooterCaptionCls = null,
                } = {}
            } = cnf,

            dimensions = {
                width,
                height,
                itemHeight, itemWidth
            },

            originalData = data.map(item => ({ [rhgID]: `${uniqueID}`, ...item })),
            innerVirtual = __getVirtual({ dimensions, size: originalData.length, scrollTop: 0, lineGap }),
            virtual = {
                loading: false,
                lineGap,
                ...innerVirtual
            },
            { fromItem, toItem } = innerVirtual,
            funcFilters = headers.reduce((acc, header) => {
                if (isFunction(header.filter)) {
                    acc[header.key] = {
                        filter: header.filter,
                        value: header.preFiltered || ''
                    };
                } else {
                    acc[header.key] = {
                        filter: () => true,
                        value: ''
                    };
                }
                return acc;
            }, {}),
            fields = headers.map(h => h.key),
            doFilter = __getDoFilter({fields, filters: funcFilters}),
            initialData = (
                globalPreFilter
                ? originalData.filter(doFilter(globalPreFilter))
                : originalData
            ).filter(doFilter());
        
        return {
            ...cnf,
            rhgID,
            originalData: originalData,
            filteredData: [...initialData],
            filtered: initialData.length,
            data: initialData.slice(fromItem, toItem),
            total: originalData.length,
            fields,
            Loader,
            header: {
                HeaderCaptionComponent,
                headerCaptionHeight
            },
            footer: {
                FooterCaptionComponent,
                footerCaptionHeight
            },
            dimensions,
            virtual,
            debounceTimes: {
                scrolling,
                filtering,
            },
            events: {
                onItemEnter,
                onItemLeave,
                onItemClick,
            },
            filters: funcFilters,
            globalFilterValue: globalPreFilter,
            cls: {
                HeaderCaptionCls,
                FooterCaptionCls
            },
            NoFilterData
            // filters: [],
            // globalFilter: ''
        };
    };

export default () => ({
    reducer,
    init
});