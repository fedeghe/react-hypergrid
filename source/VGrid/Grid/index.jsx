import React, { useCallback, useContext, useEffect, useRef } from 'react';
import NoData from './NoData';
import Filler from './Filler';
import {FILTERS} from './../constants';

import GridContext from './../Context';
import { debounce, asXsv, asJson } from './../utils';
import useStyles from './style.js';
import { ACTION_TYPES } from '../reducer';

const Grid = () => {
    const ref = useRef(),
        { state, dispatch } = useContext(GridContext),
        {
            filteredData,
            total,
            dimensions: {
                height, width,
                itemHeight, itemWidth
            },
            Item,
            virtual: {
                dataHeight,
                carpetHeight,
                contentHeight,
                scrollTop,
                loading,
                renderedItems,
                cardinality
            },
            debounceTimes: {
                scrolling: scrollingDebounceTime,
                filtering: filteringDebounceTime,
            },
            header: { HeaderCaptionComponent, headerCaptionHeight },
            footer: { FooterCaptionComponent, footerCaptionHeight },
            rhgID,
            events: {onItemEnter, onItemLeave, onItemClick },
            globalFilterValue,
            filtered,
            filters,
            columns,
            cls: { HeaderCaptionCls, FooterCaptionCls },
            filteredGroupedData: {
                allocation: { alloc, topFillerHeight, bottomFillerHeight }
            },
            grouping: {
                groupHeader : {
                    height: groupHeaderHeight,
                    Component: GroupHeaderComponent
                }
            }
        } = state,

        classes = useStyles({
            width, height,
            itemHeight, itemWidth,
            headerCaptionHeight,
            footerCaptionHeight,
        }),

        // eslint-disable-next-line react-hooks/exhaustive-deps
        globalFilter = useCallback(
            debounce(
                ({value, field}) => dispatch({
                    type: ACTION_TYPES.FILTER,
                    payload: {value, field}
                }),
                filteringDebounceTime
            ),[]
        ),

        // eslint-disable-next-line react-hooks/exhaustive-deps
        doOnScroll = useCallback(
            debounce(
                e => {
                    e.preventDefault();
                    e.stopPropagation();
                    const payload = e.target.scrollTop;
                    dispatch({
                        type: ACTION_TYPES.SCROLL,
                        payload: payload > 0 ? payload : 0
                    });
                },
                scrollingDebounceTime
            ), []
        ),

        onScroll = useCallback(e => {
            Math.abs(e.target.scrollTop - scrollTop) > (dataHeight / 4) && dispatch({
                type: ACTION_TYPES.LOADING
            });
            doOnScroll(e);
        }, [dataHeight, dispatch, doOnScroll, scrollTop]),

        getHandlers = useCallback(item => {
            const handlers = {};
            onItemEnter && (handlers.onMouseEnter = e => onItemEnter.call(e, e, {item}));
            onItemLeave && (handlers.onMouseLeave = e => onItemLeave.call(e, e, {item}));
            onItemClick && (handlers.onClick = e => onItemClick.call(e, e, {item}));
            return handlers;
        }, [onItemClick, onItemEnter, onItemLeave]),

        resetFilters = useCallback((what = FILTERS.ALL) => {
            let actionType = null;
            if (Array.isArray(what) && what.every(w => columns.includes(w))) { // is array and all in fields
                actionType = ACTION_TYPES.UNFILTER_FIELDS;
            } else if (what in FILTERS) {
                actionType = ACTION_TYPES.UNFILTER;
            }
            actionType && dispatch({
                type: actionType,
                payload: what
            });
        }, [dispatch, columns]),

        filterDataFields = useCallback(({fields}) => 
            fields
            ? filteredData.map(e => fields.reduce(
                (acc, f) => {
                    f in e && (acc[f] = e[f]);
                    return acc;
                }, {})
            )
            : filteredData
        , [filteredData]),

        downloadJson = useCallback(({fields} = {}) => {
            const a = document.createElement('a'),
                d = filterDataFields({fields}),
                blob = new Blob([JSON.stringify(asJson(d, rhgID))]);
            a.href = URL.createObjectURL(blob);
            a.target = '_blank';
            a.download = 'extract.json';                     //filename to download
            a.click();
        }, [filterDataFields, rhgID]),

        downloadXsv = useCallback(({separator = ',', fields} = {}) => {
            const a = document.createElement('a'),
                d = filterDataFields({fields}),
                xsv = asXsv((fields || columns).map(f => ({key: f})) , d, rhgID, separator),
                blob = new Blob([xsv], { type: 'text/csv' });
            a.href = URL.createObjectURL(blob);
            a.target = '_blank';
            a.download = 'extract.csv';                     //filename to download
            a.click();
        }, [columns, filterDataFields, rhgID]),

        captionProps = {
            globalFilter, 
            globalFilterValue,
            filtered,
            loading,
            renderedItems,
            total,
            cardinality,
            filters,
            resetFilters,
            downloadJson,
            downloadXsv,
            dataHeight,
            carpetHeight,
            contentHeight,
        };

    useEffect(() => {
        ref && ref.current
        && scrollTop === 0 && ref.current.scrollTo(0, 0);
    }, [scrollTop, ref]);   
    

    return <div>
        {Boolean(headerCaptionHeight) && (
            <div className={[classes.HeaderCaption, HeaderCaptionCls].join(' ')}>
                <HeaderCaptionComponent {...captionProps}/>
            </div>
        )}
        {filtered ? (
        <div className={classes.GridContainer} ref={ref} onScroll={onScroll}>
            <Filler width="100%" height={topFillerHeight} />
            {Object.entries(alloc).map(
                ([label, renderables]) => renderables.map(renderable => {
                    if (!renderable.renders) return null;
                    return renderable.header
                        ? <GroupHeaderComponent key={label} groupName={label} groupHeaderHeight={groupHeaderHeight}/>
                        : renderable.rows.map((row, i) =>
                            <div key={`${row[rhgID]}_${i}`} className={classes.Item}
                                {...getHandlers(row)}
                            >
                                <Item {...row}/>
                            </div>
                        );
                })
            )}
            <Filler width="100%" height={bottomFillerHeight} />
        </div>) : <NoData/>}
        {Boolean(footerCaptionHeight) && (
            <div className={[classes.FooterCaption, FooterCaptionCls].join(' ')}>
                <FooterCaptionComponent {...captionProps}/>
            </div>
        )}
    </div>;

};

export default Grid;