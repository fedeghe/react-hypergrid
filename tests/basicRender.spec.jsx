
/**
 * @jest-environment jsdom
 */
import React from "react";
import {
    render,
} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { configure } from "@testing-library/dom";

import VGrid from "../source/VGrid";
import { UNGROUPED_LABEL } from "../source/VGrid/constants";
import config from "./configs/basic";

configure({
    testIdAttribute: "data-uie",
});

describe("VGrid - basic", () => {
    it("should render as expected", () => {
        const { container, getByTestId, queryByTestId } = render(
            <VGrid config={config} />
        );
        let i = 0;    
        for(null; i < 20; i++)
            expect(getByTestId(`item-${UNGROUPED_LABEL}-${i}`))
        expect(container).toMatchSnapshot();

        // 20 should fail
        expect(queryByTestId(`item-${UNGROUPED_LABEL}-${i}`)).toBeNull()
    });
});
