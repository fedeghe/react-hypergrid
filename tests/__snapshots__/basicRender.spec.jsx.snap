// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VGrid - basic should render as expected 1`] = `
<div>
  <div
    class="Wrapper-0-2-1 Wrapper-d0-0-2-3 Wrapper"
  >
    <div>
      <div
        class="GridContainer-0-2-4 GridContainer-d0-0-2-8"
      >
        <div
          class="Filler-0-2-12 Filler-d0-0-2-13"
        />
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-0"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  1
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID0
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 0
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-1"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  2
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID1
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 1
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-2"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  3
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID2
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 2
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-3"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  4
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID3
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 3
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-4"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  5
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID4
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 4
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-5"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  6
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID5
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 5
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-6"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  7
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID6
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 6
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-7"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  8
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID7
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 7
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-8"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  9
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID8
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 8
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-9"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  10
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID9
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 9
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-10"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  11
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID10
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 10
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-11"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  12
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID11
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 11
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-12"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  13
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID12
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 12
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-13"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  14
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID13
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 13
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-14"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  15
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID14
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 14
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-15"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  16
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID15
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 15
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-16"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  17
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID16
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 16
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-17"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  18
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID17
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 17
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-18"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  19
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID18
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 18
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Item-0-2-5 Item-d1-0-2-9"
          data-uie="item-ungrouped-19"
        >
          <div
            class="Item-0-2-14"
          >
            <div
              class="Inner-0-2-15"
            >
              <ul>
                <li>
                  <b>
                    id
                  </b>
                   : 
                  20
                </li>
                <li>
                  <b>
                    entityid
                  </b>
                   : 
                  ID19
                </li>
                <li>
                  <b>
                    name
                  </b>
                   : 
                  a random, string 19
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="Filler-0-2-12 Filler-d1-0-2-16"
        />
      </div>
    </div>
  </div>
</div>
`;
